### PR TITLE
feat: Add support for hardhat-deploy

### LINKED ISSUE #116: Add nonce manager for quick deploys
**Is your feature request related to a problem? Please describe.**
As it stands all of our transactions are sent without waiting for receipts & without hardcoding the nonce. This means we often fail during "real" deployments (to chains other than ganache).

**Describe the solution you'd like**
Add a function `sendTransaction(tx)` which handles incrementing the nonce of each tx. This will allow us to send all of our transactions upfront & just wait until they all get mined vs waiting around for receipts.

**Describe alternatives you've considered**
Wait for receipts

### LINKED ISSUE #256: Add progress bar to contract deployment
I ran deploy, and while I see the contracts going on to the chain, I don't see anything in the console, been hanging for about 10 minutes now.

### PR DESCRIPTION
**Pull request by smartcontracts**
## Description
Adding new support for `hardhat-deploy` as an alternative to our bespoke deployment system. `hardhat-deploy` is quite nice and generates very clean artifacts. Next step should be to remove the old deployment system when possible. Though we might have to do something about generating state dumps.

fixes #256
fixes #116

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol
Comment (smartcontracts) [inline contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol:16]: This is actually very security critical. Some of our (annoying) proxies require that we call `initialize()` outside of the constructor. Making this public allows us to check that the address has been set correctly when `initialize()` is called.

File: src/hardhat-deploy-ethers.ts
Comment (smartcontracts) [inline src/hardhat-deploy-ethers.ts:52]: Call `.wait()` here so we don't have to do it every single time we want to interact with a contract.

File: tslint.json
Comment (smartcontracts) [inline tslint.json:6]: I think conditional expressions are pretty dumb

File: src/hardhat-deploy-ethers.ts
Comment (smartcontracts) [inline src/hardhat-deploy-ethers.ts:17]: Just make sure we're interacting with a contract that actually exists. Not sure if this is strictly necessary but seems better to be on the safe side.

File: deploy/Proxy__OVM_L1CrossDomainMessenger.deploy.ts
Comment (smartcontracts) [inline deploy/Proxy__OVM_L1CrossDomainMessenger.deploy.ts:43]: We can do some cool post-deploy verification checks now!

File: tslint.json
Comment (gakonst) [reply tslint.json:6]: Agree, not a fan of this syntax except for very simple cases

File: contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol
Comment (tynes) [reply contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol:16]: This needs to be in a separate PR so that this one only touches hardhat deploy related things

File: tslint.json
Comment (tynes) [reply tslint.json:6]: Can this go into `@eth-optimism/dev` and then we update the dep in another PR?

**Comment by tynes** [comment]
I'd really like it if you could add an `index.js` that does something like:

```js
const { LedgerSigner } = require('@ethersproject/hardware-wallets');

extendEnvironment((hre) => {
   const hdpath = ethers.utils.defaultPath || process.env.HD_PATH
   hre.ledger = new LedgerSigner(hre.provider, 'default', hdpath);
})
```
Ideally there is a flag `--ledger` and it just overwrites the `hre.signer` so that we don't need logic around using `hre.ledger` in each implementation of the contract deployment

**Comment by smartcontracts** [comment]
@tynes I started a conversation about hardware wallet support over here https://github.com/wighawag/hardhat-deploy/issues/88. I'm going to try experimenting with this today (assuming someone has a ledger that I can play with...)

File: contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol
Comment (smartcontracts) [reply contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol:16]: Ok I'll make a quick PR for this.

Comment (smartcontracts) [reply contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol:16]: PR is open here: https://github.com/ethereum-optimism/contracts/pull/352

Comment (smartcontracts) [reply contracts/optimistic-ethereum/libraries/resolver/Lib_AddressResolver.sol]: Done.

File: tslint.json
Comment (smartcontracts) [reply tslint.json:6]: Change to `tslint.json` was included in this PR because tslint was complaining about a file and wanted to make the file *way* uglier.

File: hardhat.config.ts
Comment (gakonst) [inline hardhat.config.ts:24]: Should probably add new mainnet/rinkeby/kovan etc. in the networks for our actual deployments.

File: src/hardhat-deploy-ethers.ts
Comment (gakonst) [inline src/hardhat-deploy-ethers.ts:28]: This is called twice in each script now, once in the external deploy script and once here.

File: src/hardhat-deploy-ethers.ts
Comment (gakonst) [inline src/hardhat-deploy-ethers.ts:82]:
```suggestion
  // Override each function call to also `.wait()` so as to simplify the deploy scripts' syntax.
  for (const fnName of Object.keys(contract.functions)) {
```

File: src/hardhat-deploy-ethers.ts
Comment (gakonst) [reply src/hardhat-deploy-ethers.ts:87]: I don't think this is a great change because it means that we'll have to "block" on the address manager registration being mined each time, resulting in slow deployments. OK for now, but let's try to remove in the future?

**Review by gakonst — DISMISSED** [review]
LGTM, some minor comments in case you want to address before merging, otherwise go for it.

**Review by maurelian — APPROVED** [review]
LGTM. 
One thing I noticed is that there are not scripts for bridge contracts. Should there be a ticket to migrate them to HH?

**Comment by snario** [comment]
@smartcontracts can you provide an update on this pull request? What's your intention for next steps? Given that it has been approved and the tests pass, I'm inclined to merge it, but I understand you've been looking at other deploy options?

**Comment by smartcontracts** [comment]
Also makes #245 obsolete

File: src/hardhat-deploy-ethers.ts
Comment (tynes) [inline src/hardhat-deploy-ethers.ts:106]: Safe code is safe

File: src/hardhat-deploy-ethers.ts
Comment (tynes) [inline src/hardhat-deploy-ethers.ts:102]: Will this allow for the `LedgerSigner`?

File: src/predeploys.ts
Comment (tynes) [inline src/predeploys.ts:1]: A nice to have imo would be to export these as contract objects that are already connected to the correct address

File: hardhat.config.ts
Comment (tynes) [inline hardhat.config.ts:59]: Why specify a key and url for each network instead of just `DEPLOYER_KEY` `RPC_URL` and `NETWORK`?

File: bin/deploy.ts
Comment (tynes) [inline bin/deploy.ts:27]: Small nit as `then` needs to be deprecated

```js
;(async () => {

})().catch(err => {

})

```

File: hardhat.config.ts
Comment (smartcontracts) [reply hardhat.config.ts:59]: good point

File: src/hardhat-deploy-ethers.ts
Comment (smartcontracts) [reply src/hardhat-deploy-ethers.ts:102]: Yep, exactly

Comment (smartcontracts) [reply src/hardhat-deploy-ethers.ts:102]: Although realistically I still think it's better to do these deployments from hot wallets and have verification once the deployment is finished, then transfer ownership over to the multisig.

File: bin/deploy.ts
Comment (smartcontracts) [reply bin/deploy.ts:27]: Was following https://hardhat.org/guides/scripts.html. And the ;(...)() syntax was giving me a lot of headache with the linter so I went with this.

File: src/predeploys.ts
Comment (smartcontracts) [reply src/predeploys.ts:1]: Yeah, will leave that for a future PR.

File: deploy/000-Lib_AddressManager.deploy.ts
Comment (ben-chain) [inline deploy/000-Lib_AddressManager.deploy.ts:39]:
```suggestion
    address: (hre as any).deployConfig.ovmProposerAddress,
```
(even if these are the same now might as well(

File: deploy/007-OVM_L1CrossDomainMessenger.deploy.ts
Comment (ben-chain) [inline deploy/007-OVM_L1CrossDomainMessenger.deploy.ts:37]: can we add a clarifying comment here that this is technically not required?  Just for auditability

File: deploy/015-OVM_L1ETHGateway.deploy.ts
Comment (ben-chain) [inline deploy/015-OVM_L1ETHGateway.deploy.ts:34]: ditto above -- comment on redundant initialization for auditability :)

**Review by ben-chain — CHANGES_REQUESTED** [review]
Requested a few nit changes, but looking great!

**Review by ben-chain — APPROVED** [review]
LGTM pending CI!

**Review by ben-chain — APPROVED** [review]
👍

**Pull request merged**
